lightning/src/ln/onion_utils.rs

```diff
@@ -506,15 +549,25 @@ pub(super) fn process_onion_failure<T: secp256k1::Signing, L: Deref>(secp_ctx: &
 	} else { unreachable!(); }
 }
 
+/// Used in the decoding of inbound payments' and onion messages' routing packets. This enum allows
+/// us to use `decode_next_hop` for returning the payloads and next hop packet bytes of both
+/// payments and onion messages.
+pub(crate) enum Payload {
+	/// This payload was for an incoming payment.
+	Payment(msgs::OnionPayload),
+	/// This payload was for an incoming onion message.
+	Message(msgs::OnionMsgPayload),
+}
+
 /// Data decrypted from the onion payload.
-pub(crate) enum Hop {
-	/// This onion payload was for us, not for forwarding to a next-hop. Contains information for
-	/// verifying the incoming payment.
-	Receive(msgs::OnionHopData),
+pub(crate) enum HopPayload {
```

Comment:
This feels super awkward. Now the caller of `decode_next_hop` (which knows the type of onion it expects) has to check that the received onion type matches what it expects? Maybe lets add trivial wrapper functions here and make `decode_next_hop` private so that the caller can still say "decode(payment_hash)" and always get back a payment onion.